pkgs/development/compilers/cudatoolkit/flags.nix

```diff
@@ -0,0 +1,69 @@
+{ config
```

Comment:
Overall, this looks great but are there docs that we can link to that explain these various settings? For example, some questions I can foresee readers asking:
* Where is the mapping from CUDA major versions to `defaultCudaCapabilities` coming from? When someone goes to add a new CUDA major version, where should they look to determine what to add to `defaultCudaCapabilities`?
* What is the significance of the "+PTX", gencode, and sm versions of capabilities?
* In `defaultCudaCapabilities`, why is `cuda10` a superset of `cuda9` but `cuda11` is not a superset of `cuda10`?
* What's the difference between an arch and a capability?

I know some of these questions may seem annoying, but easier to do O(1) work writing them down now than having to answer O(n) questions from people in the future.

Reply:
Nvidia is very flexible in terms of what flags you can use and I'm trying hard not to be prescriptive about what is best for Nix (see my note).  I've tried to keep the most common mappings Nix used already as prior art for the PR.

Docs: https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/

There is SASS code and PTX (JIT kernels).  A basic understanding can be found here: https://stackoverflow.com/questions/35656294/cuda-how-to-use-arch-and-code-and-sm-vs-compute

Most Cuda samples appear to target real architectures and a final PTX so they could for forward compatible.  That strikes me as a reasonable default.  My recommendation for a developer adding CUDA 12 would be to look at the toolkit release notes which have historically covered new and removed compute capabilities.

Library owners have it harder - they have to know how it is used and if they want SASS or PTX.  Again, I tried to match what Nix already had.

In terms of best practices for defaults, one might consider:
- Is this a supported compile architecture?  If it it is supported, is it deprecated (warnings during compile)?
- Am I looking for maximum compatibility for a wide variety of users (including older hardware)?
- Am I looking to reduce output size or compile time (reducing the options list will help)
- Do I need to provide future support for hardware that does not yet exist?
- Do I need a specific feature from a mid-generation card?
- Does my library need PTX (JIT kernels) because I am always building code it wants to compile itself

I don't think this is a one-size-fits-all.  If you're building this yourself already, you will likely want to target the specific cards you have and use.

Reply:
Thanks for this explanation! This is exactly the sort of thing that I think would be great for a comment. Even just a simple comment that links to your reply above should suffice.

Reply:
Updated docs.

Reply:
Could you also please add a code comment somewhere in this file with a copy of this explanation or with a link to your comment above? It's a great explanation and so I'd really like for it to be discoverable/persist beyond this PR review

Reply:
Makes sense, thanks for the nudge.  Done.